File: 00989. Add to Array-Form of Integer/solution.js
/**
 * Problem: 989. Add to Array-Form of Integer
 *
 * Difficulty: Easy
 *
 * Language: JavaScript
 *
 * Performance: Runtime - 0 ms (Beats 100%)
 */

/**
 * Adds integer k to the array-form of an integer num
 *
 * @param {number[]} num - Array representation of an integer
 * @param {number} k - Integer to add
 *
 * @returns {number[]} - Array representation of num + k
 */
const addToArrayForm = (num, k) => {
  // Index pointing to the rightmost digit of num
  let lastIndex = num.length - 1

  while (k) {
    // If we've processed all digits in num, add new digits at the front
    if (lastIndex < 0) num.unshift(k % 10)
    else {
      // Add k to the current digit
      k = num[lastIndex] + k
      // Store the ones digit
      num[lastIndex] = k % 10
      // Move to the next digit to the left
      lastIndex--
    }

    // Keep only the carry for the next iteration
    k = Math.floor(k / 10)
  }

  return num
}

File: 00989. Add to Array-Form of Integer/solution.go
/**
 * Problem: 989. Add to Array-Form of Integer
 *
 * Difficulty: Easy
 *
 * Language: Go
 *
 * Performance: Runtime - 0 ms (Beats 100%)
 */

func addToArrayForm(num []int, k int) []int {
	// Iterate from the end of the array while we still have a value to add
	for idx := len(num)-1; k > 0; idx-- {
		if idx >= 0 {
			// Add k to the current digit
			k += num[idx]
			// Set the current digit to the ones place of k
			num[idx] = k % 10
		} else {
			// If we've gone beyond the array, prepend the ones place of k
			num = append([]int{k % 10}, num...)
		}

		// Move to the next digit by dividing k by 10
		k /= 10
	}

	return num
}